diag_manager/fms_diag_object_container.F90

```diff
@@ -0,0 +1,201 @@
+
+
+contains
+
+   !> \brief Returns an empty itetator if a diag object with this ID was not found.
+   ! If the diag object was found, return an iterator with the current object being
+   ! the found object, ad the last/anchor being the last/anchor of the container.
+   ! Note that this routine can accept an optional iterator as input, which
+   ! is useful for chaining searches, which may be needed if there are key duplicates.
+   function find_diag_object (this, id , iiter) result (riter)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer , intent (in) :: id
+      class(diag_obj_iterator), intent (in), optional :: iiter
+      class(diag_obj_iterator) , allocatable :: riter
+      !!
+      class(fms_diag_object),  allocatable , target:: tdo !!a temporaty diagnostic object
+      integer :: status
+      !!
+      if(present (iiter)) then
+         riter = iiter
+      else
+         riter = this%iterator() !!An iterator over the entire container
+      endif
+      do while( riter%has_data() .eqv. .true.)
+         tdo = riter%get()
+         if(id == tdo%get_id() ) then
+            EXIT
+         end if
+         status = riter%next()
+      end do
+   end function find_diag_object
+
+   !> \brief insert diagnostic object obj with given id.
+   ! Objects are inserted at the back / end of the list
+   ! This version of the container also enforces that the
+   ! objects ID is equal the input key. status of -1
+   ! is returned fro errors
+   !! TODO: what should be the reutrn type
+   !!TODO: WE ASSUME id == obj%diagID
+   function insert_diag_object (this, id, obj) result (status)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer,  intent (in) :: id
+      class(fms_diag_object) , intent (in out) :: obj
+      integer :: status
+      !!
+      class(literator), allocatable ::  tliter
+
+      status = -1
+      if ( id .ne. obj%get_id() ) then
+         !!TODO: log error
+      endif
+      tliter =  this%the_ll%push_back( obj )
+      if(tliter%has_data() .eqv. .true. ) then
+         status = 0
+      endif
+   end function
+
+   !> \brief Remove and return the first object in the container with the corresponding key .
+   ! Return an iterator that starts with the next object on the container. (Note, if the client
+   ! code does not already have a referenc to the object being removed, then  the client may
+   ! want to to use subroutine find before using subroutine remove. If subroutine find is used,
+   ! consider then calling remove with the iterator returned from find.)
+   function remove_diag_object (this, key, iiter ) result (riter)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer , intent (in) :: key
+      class(diag_obj_iterator), intent (in), optional :: iiter
+      class(diag_obj_iterator), allocatable:: riter
+      class(literator), allocatable :: temp_liter
+      integer :: status
```

Comment:
Variables need to be documented in doxygen style

Reply:
Done